# Heart Disease Framingham

## Introduction




### Problem

The World Health Organization has estimated 12 million deaths occur worldwide, every year due to Heart diseases. Half the deaths in the United States and other developed countries are due to cardio vascular diseases. The early prognosis of cardiovascular diseases can aid in making decisions on lifestyle changes in high risk patients and in turn reduce the complications. This project will compare the performance of different models and try to improve the predictive performance.

### Source 

The dataset is publically available on the Kaggle website, and it is from an ongoing cardiovascular study on residents of the town of Framingham, Massachusetts. The classification goal is to predict whether the patient has 10-year risk of future coronary heart disease (CHD).The dataset provides the patients’ information. It includes over 4,000 records and 15 attributes. Variables Each attribute is a potential risk factor. There are both demographic, behavioral and medical risk factors.

## Logistic Regression Model

In [ ]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/soltaniehha/Intro-to-Data-Analytics/master/data/AnalyticsEdge-Datasets/Framingham.csv')
df.head(3)

male  age  education  currentSmoker  ...    BMI  heartRate  glucose  TenYearCHD
0     1   39        4.0              0  ...  26.97       80.0     77.0           0
1     0   46        2.0              0  ...  28.73       95.0     76.0           0
2     1   48        1.0              1  ...  25.34       75.0     70.0           0

[3 rows x 16 columns]

### Data Description

* Demographic risk factors
  * male: sex of patient
  * age: age in years at first examination
  * education: Some high school (1), high school/GED (2), some college/vocational school (3), college (4)
* Behavioral risk factors
  * currentSmoker, cigsPerDay: Smoking behavior
  * Medical history risk factors
  * BPmeds: On blood pressure medication at time of first examination
  * prevalentStroke: Previously had a stroke
  * prevalentHyp: Currently hypertensive
  * diabetes: Currently has diabetes
* Risk factors from first examination
  * totChol: Total cholesterol (mg/dL)
  * sysBP: Systolic blood pressure
  * diaBP: Diastolic blood pressure
  * BMI: Body Mass Index, weight (kg)/height (m)2
  * heartRate: Heart rate (beats/minute)
  * glucose: Blood glucose level (mg/dL)



### Preprocessing - no predictive power/handling missing values

In [ ]:
# Overall the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4240 entries, 0 to 4239
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   male             4240 non-null   int64  
 1   age              4240 non-null   int64  
 2   education        4135 non-null   float64
 3   currentSmoker    4240 non-null   int64  
 4   cigsPerDay       4211 non-null   float64
 5   BPMeds           4187 non-null   float64
 6   prevalentStroke  4240 non-null   int64  
 7   prevalentHyp     4240 non-null   int64  
 8   diabetes         4240 non-null   int64  
 9   totChol          4190 non-null   float64
 10  sysBP            4240 non-null   float64
 11  diaBP            4240 non-null   float64
 12  BMI              4221 non-null   float64
 13  heartRate        4239 non-null   float64
 14  glucose          3852 non-null   float64
 15  TenYearCHD       4240 non-null   int64  
dtypes: float64(9), int64(7)
memory usage: 530.1 KB


In [ ]:
# Check missing values
df.isnull().sum()

male                 0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

First, we will drop the `education` variable. Because all people at any level of education are equal in the face of disease, `education` should not be used as a variable affecting disease. Meanwhile, for the medical industry, subjective variables such as education level should not appear in the results of machine learning, which may affect doctors' judgment of patients.

In [ ]:
# Drop education column
df.drop(['education'], axis=1, inplace=True)    # drop deck, embark_town, alive, class & sex

Among the remaining variables, `cigsPerDay`, `BPMeds`, `totChol`, `BMI` and `glucose` still have null values. These null values account for about 12% of the total data, given the medical importance and small proportion of these variables. Dropping these entries does not lose too much data.

In [ ]:
# Calculate missing entries percentage
missing_values = df.isnull().sum().sort_values(ascending=False)
missing_percentage = (df.isnull().sum()/df.isnull().count()).sort_values(ascending=False)*100
missing_results = pd.concat([missing_values, missing_percentage], axis=1, keys=["Total", "Percentage"])
missing_data = missing_results[missing_results['Total']>0]
missing_data

Total  Percentage
glucose       388    9.150943
BPMeds         53    1.250000
totChol        50    1.179245
cigsPerDay     29    0.683962
BMI            19    0.448113
heartRate       1    0.023585

In [ ]:
# Drop missing entries
df.dropna(axis=0, inplace=True)
df.isnull().sum()

male               0
age                0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
TenYearCHD         0
dtype: int64

Let's check the remaining amount of data, we still have 3751 rows of data for further analysis.

In [ ]:
# Check remain entires
df.shape

(3751, 15)

### Preprocessing - categorical variables
Before machine learning, we need to make sure that all the variables are numerical variables, so it looks like our data is clean and we don't need to do any more processing.

In [ ]:
# Show dataset sample
df.head(3)

male  age  currentSmoker  cigsPerDay  ...    BMI  heartRate  glucose  TenYearCHD
0     1   39              0         0.0  ...  26.97       80.0     77.0           0
1     0   46              0         0.0  ...  28.73       95.0     76.0           0
2     1   48              1        20.0  ...  25.34       75.0     70.0           0

[3 rows x 15 columns]

### Create feature matrix and target vector

In [ ]:
# Create a feature matrix without `TenYearCHD` and call it `X`:
X = df.drop('TenYearCHD', axis=1)
X.shape

(3751, 14)

In [ ]:
# Create a target vector with "survived" and call it `y`:
y = df['TenYearCHD']
y.shape

(3751,)

### Split the data randomly into train and test

Next, the data will be split as test and train dataset. Test data accounted for 30% of the metadata set, and training data accounted for 70% of the metadata set.

In [ ]:
# Split dataset randomly by using sklearn package
from sklearn.model_selection import train_test_split
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.3, random_state=780)

In [ ]:
Xtrain.shape

(2625, 14)

In [ ]:
ytrain.shape

(2625,)

In [ ]:
Xtest.shape

(1126, 14)

In [ ]:
ytest.shape

(1126,)

### Instantiate and fit a logistic regression model

In [ ]:
# Set "liblinear" as solver for logistic regression
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='liblinear')

In [ ]:
# Fit model to the training data
model.fit(Xtrain, ytrain)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

### Prediction

In [ ]:
# predict on test data and store the results as y_model
y_model = model.predict(Xtest)

In [ ]:
# Add predit results to test dataset
test = Xtest.join(ytest).reset_index()
test.join(pd.Series(y_model, name='predicted')).head()

index  male  age  currentSmoker  ...  heartRate  glucose  TenYearCHD  predicted
0   1290     0   40              1  ...       88.0     67.0           0          0
1   2025     0   57              0  ...       83.0     75.0           1          0
2   1477     0   49              0  ...       80.0     74.0           0          0
3    603     0   61              1  ...       90.0     91.0           1          0
4   2380     1   59              1  ...       54.0     85.0           0          0

[5 rows x 17 columns]

#### Accuracy

Our basic model is giving us an accuracy of 84.6%, which is a very good prediction result.

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(ytest, y_model)

0.8463587921847247

#### Sensitivity
Sensitivity = $TPR=\frac{TP}{P}$, where $TP$ is ture positives and $P$ is count of all positives.

In [ ]:
P = sum(ytest == 1)
TP = sum((ytest == 1) & (y_model == 1))
TP/P

0.022988505747126436

#### Specificity
Specificity = $TNR=\frac{TN}{N}$, where $TN$ is ture negatives and $N$ is count of all negatives.

In [ ]:
N = sum(ytest == 0)
TN = sum((ytest == 0) & (y_model == 0))
TN/N

0.9968487394957983

In [ ]:
# Calculating the mean absolute error
from sklearn.metrics import mean_absolute_error
mean_absolute_error(ytest, y_model)

0.15364120781527532

## Performance Improvement

### Feature Engineering

First, have a look at the correlation matrix between Predictor variable and target variable.

In [ ]:
df.corr()

male       age  ...   glucose  TenYearCHD
male             1.000000 -0.024142  ...  0.003130    0.096000
age             -0.024142  1.000000  ...  0.118529    0.231584
currentSmoker    0.203215 -0.211338  ... -0.054078    0.021709
cigsPerDay       0.325886 -0.188804  ... -0.055090    0.055920
BPMeds          -0.052355  0.131690  ...  0.052464    0.084740
prevalentStroke -0.002513  0.050018  ...  0.015789    0.047684
prevalentHyp     0.003588  0.305292  ...  0.085776    0.178615
diabetes         0.011834  0.109321  ...  0.616087    0.093222
totChol         -0.067595  0.261443  ...  0.046902    0.089613
sysBP           -0.044515  0.388280  ...  0.132878    0.220108
diaBP            0.053890  0.205191  ...  0.061731    0.148994
BMI              0.074868  0.136428  ...  0.088133    0.084400
heartRate       -0.115444 -0.005893  ...  0.099575    0.022658
glucose          0.003130  0.118529  ...  1.000000    0.124117
TenYearCHD       0.096000  0.231584  ...  0.124117    1.000000

[15 rows x 15 columns]

According to the correlation matrix, it can be seen that there are some variables with low values. In the following steps, these variables will be further processed to improve their correlation and thus improve the performance of the model.
* **male**: sex of patient
* **currentSmoker**: Smoking behavior
* **cigsPerDay**: Smoking behavior
* **BPMeds**: On blood pressure medication at time of first examination
* **prevalentStroke**: Previously had a stroke
* **diabetes**: Currently has diabetes
* **totChol**: Total cholesterol (mg/dL)
* **BMI**: Body Mass Index, weight (kg)/height (m)2
* **heartRate**: Heart rate (beats/minute)

In [ ]:
df_eng = df.copy()

#### Create BMI Level variable
Source: [American Heart Association - BMI level](https://www.heart.org/en/healthy-living/healthy-eating/losing-weight/bmi-in-adults)

According to the classification of BMI by American Heart Association, I will classify these variables and generate a new variable: BMI_level.
* Underweight: BMI <= 18.5 kg/m²
* Healthy: 18.5 kg/m² < BMI < 24.9 kg/m²
* Overweight: 25 kg/m² <= BMI < 30.0 kg/m²
* Obesity: BMI >= 30.0 kg/m²
* Extreme obesity: BMI >= 40 kg/m²

In [ ]:
import numpy as np

# create a list of our conditions
conditions = [
    (df_eng['BMI'] <= 18.5),
    (df_eng['BMI'] > 18.5) & (df_eng['BMI'] < 25),
    (df_eng['BMI'] >= 25) & (df_eng['BMI'] < 30),
    (df_eng['BMI'] >= 30) & (df_eng['BMI'] < 40),
    (df_eng['BMI'] >= 40)
    ]

# create a list of the values we want to assign for each condition
values = ['Underweight', 'Healthy', 'Overweight', 'Obesity', 'Extreme obesity']

# create a new column and use np.select to assign values to it using our lists as arguments
df_eng['BMI_level'] = np.select(conditions, values)

# display updated DataFrame
df.head()

male  age  currentSmoker  cigsPerDay  ...    BMI  heartRate  glucose  TenYearCHD
0     1   39              0         0.0  ...  26.97       80.0     77.0           0
1     0   46              0         0.0  ...  28.73       95.0     76.0           0
2     1   48              1        20.0  ...  25.34       75.0     70.0           0
3     0   61              1        30.0  ...  28.58       65.0    103.0           1
4     0   46              1        23.0  ...  23.10       85.0     85.0           0

[5 rows x 15 columns]

#### Create totChol Level variable
Source: [Medical News Today - Cholesterol range](https://www.medicalnewstoday.com/articles/315900)

According to the classification of Total Cholesterol by Medical News Today, I will classify these variables and generate a new variable: totChol_range.
* Desirable: totChol <= 200 mg/dL
* Borderline High: 200 mg/dL < totChol < 240 mg/dL
* High: totChol >= 240 mg/dL

In [ ]:
import numpy as np

# create a list of our conditions
conditions = [
    (df_eng['totChol'] <= 200),
    (df_eng['totChol'] > 200) & (df_eng['totChol'] < 240),
    (df_eng['totChol'] >= 240)
    ]

# create a list of the values we want to assign for each condition
values = ['Desirable', 'Borderline High', 'High']

# create a new column and use np.select to assign values to it using our lists as arguments
df_eng['totChol_level'] = np.select(conditions, values)

# display updated DataFrame
df.head()

male  age  currentSmoker  cigsPerDay  ...    BMI  heartRate  glucose  TenYearCHD
0     1   39              0         0.0  ...  26.97       80.0     77.0           0
1     0   46              0         0.0  ...  28.73       95.0     76.0           0
2     1   48              1        20.0  ...  25.34       75.0     70.0           0
3     0   61              1        30.0  ...  28.58       65.0    103.0           1
4     0   46              1        23.0  ...  23.10       85.0     85.0           0

[5 rows x 15 columns]

#### Create BP (Blood Pressure) Level variable
Source: [American Heart Association - Blood Pressure range](https://www.heart.org/en/health-topics/high-blood-pressure/understanding-blood-pressure-readings)

According to the classification of Blood Pressure by American Heart Association, I will classify these variables and generate a new variable: BP_level.
* Normal: sysBP <= 120 mmHg && diaBP <= 80 mmHg
* Elevated: 120 mmHg < sysBP < 130 mmHg && diaBP <= 80 mmHg
* High blood pressure 1: 130 mmHg <= sysBP < 140 mmHg || 80 mmHg < diaBP < 90 mmHg
* High blood pressure 2: sysBP >= 140 mmHg || diaBP >= 90 mmHg
* Hypertensive Crisis: sysBP >= 180 mmHg && diaBP >= 120 mmHg

In [ ]:
import numpy as np

# create a list of our conditions
conditions = [
    (df_eng['sysBP'] <= 120) & (df_eng['diaBP'] <= 80),
    ((df_eng['sysBP'] > 120) & (df_eng['sysBP'] < 130)) & (df_eng['diaBP'] <= 80),
    ((df_eng['sysBP'] >= 130) & (df_eng['sysBP'] < 140)) | ((df_eng['diaBP'] > 80) & (df_eng['diaBP'] <= 90)),
    (df_eng['sysBP'] >= 140) & (df_eng['diaBP'] >= 90),
    (df_eng['sysBP'] >= 180) & (df_eng['diaBP'] >= 120)
    ]

# create a list of the values we want to assign for each condition
values = ['Normal', 'Elevated', 'High blood pressure 1', 'High blood pressure 2', 'Hypertensive Crisis']

# create a new column and use np.select to assign values to it using our lists as arguments
df_eng['BP_level'] = np.select(conditions, values)

# display updated DataFrame
df.head()

male  age  currentSmoker  cigsPerDay  ...    BMI  heartRate  glucose  TenYearCHD
0     1   39              0         0.0  ...  26.97       80.0     77.0           0
1     0   46              0         0.0  ...  28.73       95.0     76.0           0
2     1   48              1        20.0  ...  25.34       75.0     70.0           0
3     0   61              1        30.0  ...  28.58       65.0    103.0           1
4     0   46              1        23.0  ...  23.10       85.0     85.0           0

[5 rows x 15 columns]

#### Preprocessing - categorical variables

After adding new variables, there are some categorical variables, which we will handle next.

In [ ]:
df_eng = pd.get_dummies(df_eng, columns=['BMI_level', 'totChol_level', 'BP_level'])
df_eng.head()

male  age  ...  BP_level_High blood pressure 2  BP_level_Normal
0     1   39  ...                               0                1
1     0   46  ...                               0                0
2     1   48  ...                               0                0
3     0   61  ...                               1                0
4     0   46  ...                               0                0

[5 rows x 28 columns]

Now let's see if the correlation matrix is improved after adding new variables。

In [ ]:
df_eng.corr()

male  ...  BP_level_Normal
male                            1.000000  ...        -0.048519
age                            -0.024142  ...        -0.251690
currentSmoker                   0.203215  ...         0.110724
cigsPerDay                      0.325886  ...         0.056012
BPMeds                         -0.052355  ...        -0.106779
prevalentStroke                -0.002513  ...        -0.030795
prevalentHyp                    0.003588  ...        -0.407692
diabetes                        0.011834  ...        -0.045656
totChol                        -0.067595  ...        -0.174540
sysBP                          -0.044515  ...        -0.617992
diaBP                           0.053890  ...        -0.600470
BMI                             0.074868  ...        -0.260508
heartRate                      -0.115444  ...        -0.139750
glucose                         0.003130  ...        -0.069874
TenYearCHD                      0.096000  ...        -0.117033
BMI_level_Extreme obesity      -0.047733  ...        -0.047827
BMI_level_Healthy              -0.141340  ...         0.226882
BMI_level_Obesity              -0.023459  ...        -0.141052
BMI_level_Overweight            0.177342  ...        -0.143543
BMI_level_Underweight          -0.052671  ...         0.067508
totChol_level_Borderline High   0.076178  ...         0.035554
totChol_level_Desirable        -0.003725  ...         0.144159
totChol_level_High             -0.069445  ...        -0.151864
BP_level_0                     -0.011863  ...        -0.123512
BP_level_Elevated              -0.014136  ...        -0.205367
BP_level_High blood pressure 1  0.055701  ...        -0.512257
BP_level_High blood pressure 2  0.002367  ...        -0.292740
BP_level_Normal                -0.048519  ...         1.000000

[28 rows x 28 columns]

In [ ]:
# Create a feature matrix without `TenYearCHD` and call it `X`:
X_eng = df_eng.drop('TenYearCHD', axis=1)
# Create a target vector with "survived" and call it `y`:
y_eng = df_eng['TenYearCHD']
# Split dataset randomly by using sklearn package
from sklearn.model_selection import train_test_split
Xtrain_eng, Xtest_eng, ytrain_eng, ytest_eng = train_test_split(X_eng, y_eng, test_size=0.3, random_state=780)
# Set "liblinear" as solver for logistic regression
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='liblinear')
# Fit model to the training data
model.fit(Xtrain_eng, ytrain_eng)
# predict on test data and store the results as y_model
y_model_eng = model.predict(Xtest_eng)
# Add predit results to test dataset
test_eng = Xtest_eng.join(ytest_eng).reset_index()
test_eng.join(pd.Series(y_model_eng, name='predicted')).head()

index  male  age  ...  BP_level_Normal  TenYearCHD  predicted
0   1290     0   40  ...                1           0          0
1   2025     0   57  ...                0           1          0
2   1477     0   49  ...                0           0          0
3    603     0   61  ...                0           1          0
4   2380     1   59  ...                0           0          0

[5 rows x 30 columns]

#### Improved model's accuracy, sensitivity and specificity


By adding new variables, accuracy, sensitivity, mean absolute error and specificity were improved by 1%, 3% , 0.1%, 1%, which were huge improvement.

In [ ]:
# Calculate accuracy
from sklearn.metrics import accuracy_score
accuracy_score(ytest_eng, y_model_eng)

0.8516873889875666

In [ ]:
# Calculate sensitivity
P = sum(ytest_eng == 1)
TP = sum((ytest_eng == 1) & (y_model_eng == 1))
TP/P

0.05172413793103448

In [ ]:
# Calculate specificity
N = sum(ytest_eng == 0)
TN = sum((ytest_eng == 0) & (y_model_eng == 0))
TN/N

0.9978991596638656

In [ ]:
# Calculating the mean absolute error
from sklearn.metrics import mean_absolute_error
mean_absolute_error(ytest_eng, y_model_eng)

0.1483126110124334


At the same time, I tried to delete BMI, totChol, sysBP and diaBP, because the variables we created were derived from the data of these variables. However, I found that if these variables were deleted, the performance of the model would be limited, with accuracy, sensitivity increasing by 0.5%, 2%.

### Try different model: Random Forest

#### Original database: df

Firstly, random Forest model was used for the dataset with no new variables added.

The original performance is:
* Accuracy: 0.8463587921847247
* Sensitivity: 0.022988505747126436
* Specificity: 0.9968487394957983
* Mean absolute error: 0.15364120781527532


In [ ]:
# Run Random Forest model
from sklearn.ensemble import RandomForestClassifier
regressor = RandomForestClassifier(n_estimators=20, random_state=0)
regressor.fit(Xtrain, ytrain)
y_model = regressor.predict(Xtest)

In [ ]:
# Calculate accuracy
from sklearn.metrics import accuracy_score
accuracy_score(ytest, y_model)

0.8481349911190054

In [ ]:
# Calculate sensitivity
P = sum(ytest == 1)
TP = sum((ytest == 1) & (y_model == 1))
TP/P

0.07471264367816093

In [ ]:
# Calculate specificity
N = sum(ytest == 0)
TN = sum((ytest == 0) & (y_model == 0))
TN/N

0.9894957983193278

In [ ]:
# Calculating the mean absolute error
from sklearn.metrics import mean_absolute_error
mean_absolute_error(ytest, y_model)

0.15186500888099466

Compared with previous logistic regression results, accuracy, sensitivity and specificity were reduced by 0.2%, 5% and 0.5%. Although accuracy has decreased, mean absolute error has decreased by 0.2%.

In general, this is a good improvement, because the sensitivity has been improved a lot.

#### Improved database: df_eng

Secondly, random Forest model was used for the dataset with new variables added.

The original performance is:
* Accuracy: 0.8516873889875666
* Sensitivity: 0.05172413793103448
* Specificity: 0.9978991596638656
* Mean absolute error: 0.1483126110124334


In [ ]:
# Run Random Forest model
from sklearn.ensemble import RandomForestClassifier
regressor = RandomForestClassifier(n_estimators=20, random_state=0)
regressor.fit(Xtrain_eng, ytrain_eng)
y_model_eng = regressor.predict(Xtest_eng)

In [ ]:
# Calculate accuracy
from sklearn.metrics import accuracy_score
accuracy_score(ytest_eng, y_model_eng)

0.8374777975133215

In [ ]:
# Calculate sensitivity
P = sum(ytest_eng == 1)
TP = sum((ytest_eng == 1) & (y_model_eng == 1))
TP/P

0.05747126436781609

In [ ]:
# Calculate specificity
N = sum(ytest_eng == 0)
TN = sum((ytest_eng == 0) & (y_model_eng == 0))
TN/N

0.9800420168067226

In [ ]:
# Calculating the mean absolute error
from sklearn.metrics import mean_absolute_error
mean_absolute_error(ytest_eng, y_model_eng)

0.1625222024866785

It seems that random forest is not a suitable model for new data. Accuracy, sensitivity, specificity and mean absolute error were not enhanced, but decreased.

## Conclusion

By comparing the models, I found that the performance of a dataset varies from model to model. 

For example, in Logistic Regression, we added new variables to improve performance. However, using this data set in the Random Forest model does not perform as well as the data set without adding data.

Therefore, it is not necessary to increase or decrease variables to improve model performance, which is relatively speaking. There is a suitable model for each data set. Therefore, when selecting models and data sets, performance is one of the factors to be considered, and the number of data variables should also be considered to avoid overfitting.